rero_ils/modules/documents/dojson/contrib/jsontomarc21/model.py

```diff
@@ -0,0 +1,507 @@
+    for source in source_order:
+        source_data = contribution.get('agent', {}).get(source, {})
+        if source_data:
+            refs.append({
+                'source': source,
+                'pid': source_data.get("pid")
+            })
+            agent = set_value(
+                source_data, agent, 'bf:Agent')
+            agent = set_value(
+                source_data, agent, 'preferred_name')
+            agent = set_value(
+                source_data, agent, 'numeration')
+            agent = set_value(
+                source_data, agent, 'qualifier')
+            agent = set_value(
+                source_data, agent, 'date_of_birth')
+            agent = set_value(
+                source_data, agent, 'date_of_death')
+            agent = set_value(
+                source_data, agent, 'subordinate_unit')
+            agent = set_value(
+                source_data, agent, 'conference')
+            agent = set_value(
+                source_data, agent, 'conference_number')
+            agent = set_value(
+                source_data, agent, 'conference_date')
+            agent = set_value(
+                source_data, agent, 'conference_place')
+            agent.pop(source)
+    agent['refs'] = refs
+    agent['type'] = agent.pop('bf:Agent')
+    contribution['agent'] = agent
+    return contribution
+
+
+def get_holdings_items(document_pid):
+    """Create Holding and Item informations.
+
+    :param document_pid: document pid to use for holdings search
+    :param organisations_pid: organisation pid to use for holdings search
+    """
+    from rero_ils.modules.holdings.api import Holding, HoldingsSearch
+    from rero_ils.modules.items.api import Item, ItemsSearch
+    from rero_ils.modules.libraries.api import Library
+    from rero_ils.modules.locations.api import Location
+    from rero_ils.modules.organisations.api import Organisation
+
+    @memoized()
+    def get_name(resource, pid):
+        """Get name from resource."""
+        data = resource.get_record_by_pid(pid)
+        if data:
+            return data.get('name')
+
+    results = []
+    if document_pid:
+        holding_pids = Holding.get_holdings_pid_by_document_pid(
+            document_pid=document_pid,
+            with_masked=False
+        )
+
+        holding_pids = list(holding_pids)
+        hits = HoldingsSearch().filter('terms', pid=holding_pids).scan()
+        for hit in hits:
+            holding = hit.to_dict()
+            result = {
+                'organisation': {
+                    'pid': hit.organisation.pid,
+                    'name': get_name(Organisation, hit.organisation.pid)
+                },
+                'library': {
+                    'pid': hit.library.pid,
+                    'name': get_name(Library, hit.library.pid)
+                },
+                'location': {
+                    'pid': hit.location.pid,
+                    'name': get_name(Location, hit.location.pid)
+                },
```

Comment:
For each holding hit, you reload the organisation, the library, the location. But these informations should be load once and temporary store to improve speed. Something like :
```python
orgs = libraries = locations = {}
for hit in hits:
    holding = hit.to_dict()
    if hit.organisation.pid not in orgs:
        orgs[hit.organisation.pid] = get_name(Organisation, hit.organisation.pid)
    ...
    result = {
        'organisation': { 'pid': hit.organisation.pid, 'name': orgs[hit.organisation.pid]},
         ...
     }
     
```